cf_cli_java_plugin.go

```diff
@@ -229,7 +249,29 @@ func (c *JavaPlugin) execute(commandExecutor cmd.CommandExecutor, uuidGenerator
 	cfSSHArguments = append(cfSSHArguments, remoteCommand)
 
 	output, err := commandExecutor.Execute(cfSSHArguments)
+	if command == heapDumpCommand {
+
+		finalFile, err := util.FindDumpFile(applicationName, heapdumpFileName)
+		if err != nil && finalFile != "" {
+			heapdumpFileName = finalFile
+		}
 
+		if copyToLocal {
+			err = util.CopyOverCat(applicationName, heapdumpFileName, localFileFullPath)
+			if err == nil {
+				fmt.Println("heap dump filed saved to: " + localFileFullPath)
+			} else {
+				return "", err
+			}
+		}
+
+		if !keepAfterDownload {
+			err = util.DeleteRemoteFile(applicationName, heapdumpFileName)
+			if err != nil {
+				return "", err
```

Comment:
Maybe it's worth adding:

```golang
fmt.Println("heap dump filed deleted in app container")
```

Reply:
> I've added some more comments to go through.
> In general, we're missing to test the whole `utils/cfutils.go`. However, since I know that you're currently assigned with limited time, we can take this as a follow up improvement.
> 
> Before approving I'd like to quickly try out this build of the cf plugin myself with a demo Java app. I just saw that we have never added to the README how one can test the plugin locally. Can you maybe share how you've tested the plugin so that I do the test myself?

Summarized test cases from my side:
java-app: spring-music
file system mounted path: /var/fsjavadev

test case 1 - without any parameters set: 
command: cf java heap-dump spring-music
output: 
    heap dump will not be copied as parameter `local-dir` was not set
    heap dump filed deleted in app container
behind the ouput: monitor the container by ssh command 'cf ssh spring-music', will find during the plugin running, there will be one dump file with name 'spring-music-heapdump-<guid>.hprof' generated in /var/fsjavadev, and removed when the plugin exit running.

test case 2 - with parameter keep set:
command: cf java heap-dump spring-music -k
output:
    heap dump will not be copied as parameter `local-dir` was not set
behind the ouput: dump file should be kept in container under directory: /var/fsjavadev, we can check with command: 'cf ssh spring-music -c "ls -lrt /var/fsjavadev/*hprof"' and got result: 
-rw------- 1 vcap vcap 62270162 Aug 11 02:50 /var/fsjavadev/spring-music-heapdump-b9faef01-f061-4d9b-a0d0-785f741e4205.hprof

test case 3 - with invalid container dir set:
command: cf java heap-dump spring-music -container-dir /var/notexists
output:
    FAILED
    the container path specified doesn't exist or have no read and write access, please check and try again later

test case 4 - with valid container dir set:
command: cf java heap-dump spring-music -container-dir /var/fsjavadev
    heap dump will not be copied as parameter `local-dir` was not set
    heap dump filed deleted in app container
behind the ouput: monitor the container by ssh command 'cf ssh spring-music', will find during the plugin running, there will be one dump file with name 'spring-music-heapdump-<guid>.hprof' generated in /var/fsjavadev, and removed when the plugin exit running.

test case 5 - with valid container dir and keep set:
command: cf java heap-dump spring-music -container-dir /var/fsjavadev -keep
    heap dump will not be copied as parameter `local-dir` was not set
behind the ouput: dump file should be kept in container under directory: /var/fsjavadev, we can check with command: 'cf ssh spring-music -c "ls -lrt /var/fsjavadev/*hprof"' and got result: 
-rw------- 1 vcap vcap 62270162 Aug 11 02:50 /var/fsjavadev/spring-music-heapdump-b9faef01-f061-4d9b-a0d0-785f741e4205.hprof <--from test case 2
-rw------- 1 vcap vcap 62662145 Aug 11 02:59 /var/fsjavadev/spring-music-heapdump-9fd5d28c-3a60-4eda-8998-2f794d9fce29.hprof

test case 6 - with valid container dir and local-dir set:
command: cf java heap-dump spring-music -container-dir /tmp -local-dir /Users/i318048
    heap dump filed saved to: /Users/i318048/spring-music-heapdump-492b171a-5969-4b53-ae26-f4f0b2770c2c.hprof
    heap dump filed deleted in app container
behind the output: dump file was generated in /tmp and get deleted when plugin exits running

test case 7 - with valid container dir and invalid local-dir set:
command: cf java heap-dump spring-music -container-dir /tmp -local-dir /Users/invalid
output:
    FAILED
    Error creating local file at  /Users/invalid/spring-music-heapdump-91b34f7e-a3fd-42b0-af73-eb1f07f714e6.hprof. Please check that you are allowed to create files at the given local path.
behind the output: the dump file was generated under specified path /tmp, and user can copy the via a command like 'cf ssh spring-music -c "cat /tmp/spring-music-heapdump-<guid>.hprof" > /valid/local/path/file_name '


test case 8 - with local-dir set:
command: cf java heap-dump spring-music -local-dir /Users/i318048
output:
    heap dump filed saved to: /Users/i318048/spring-music-heapdump-860f7175-e42d-4f40-b7ee-9a651946735b.hprof
    heap dump filed deleted in app container

part 2, required tools not ready: 
java-app: firstappzw
file system: not present

test case 1: ssh disabled
command: cf java heap-dump firstappzw  -k 
output: 
    FAILED
    ssh is not enabled for app: 'firstappzw', please run below 2 shell commands to enable ssh and try again(please note application should be restarted before take effect):
    cf enable-ssh firstappzw
    cf restart firstappzw

test case 2: jvmmon and jmap missing(currently remove these two tools from container):
command: cf java heap-dump firstappzw  -k
output:
    FAILED
    jvmmon or jmap are required for generating heap dump, you can modify your application manifest.yaml on the 'JBP_CONFIG_OPEN_JDK_JRE' environment variable. This could be done like this:
                    ---
                    applications:
                    - name: <APP_NAME>
                    memory: 1G
                    path: <PATH_TO_BUILD_ARTIFACT>
                    buildpack: https://github.com/cloudfoundry/java-buildpack
                    env:
                            JBP_CONFIG_OPEN_JDK_JRE: '{ jre: { repository_root: "https://java-buildpack.cloudfoundry.org/openjdk-jdk/bionic/x86_64", version: 11.+ } }'

test case 3: with valid container dir and local dir set:
command: cf java heap-dump firstappzw -container-dir /tmp -local-dir /Users/i318048
output:
    heap dump filed saved to: /Users/i318048/firstappzw-heapdump-923e9722-423c-4724-702f-3e2b17132b5a.hprof
    heap dump filed deleted in app container
behind the output: dump file was generated in /tmp and get deleted when plugin exits running

test case 4: with valid local dir set:
command: cf java heap-dump firstappzw -local-dir /Users/i318048
output:
    heap dump filed saved to: /Users/i318048/firstappzw-heapdump-923e9722-423c-4724-702f-3e2b17132b5a.hprof
    heap dump filed deleted in app container
behind the output: without container directory specified, the dump file will be saved to file system attached, and to /tmp if file system not present

Reply:
output message after deleted dump file.